### PR TITLE
Development updates merge to master

### PR DESCRIPTION
**Pull request by bernardkwok**
*Development updates*
Implementations:
- fix GLSL rgbtohsv and hsvtorgb to use OSL logic to incorrect logic.
- fix GLSL smoothstep, burn, dodge, disjoint over to handle division by 0 out of range fallbacks.
- make burley-diffuse OSL a no-op for now as oren_nayer built-in call fails.

ShaderGen:
- Fix CM transform insertion between nodes + add test case.

Unit tests:
- Add testing of object vs world space for positon and normal
- Add unit tests for complete coverage of 2D and 3D noise.
- Update geometry and camera placement to have matching OSL/GLSL input scenes.
- Make the test geometry uvs match the OSL geometry for testrender
  - add in MeshStream transform function.
- Replace placement texture (marble) with new "grid" texture.
- Add in Python script utility to create web page from rendered results.
- Migrate so _options.mtlx can be used for Gen* tests. Gen* tests now check for coverage of implementation usage.

Viewer:
- Add in env display option.
- Unshare trianges to handle uv borders. Was previously sharing all uvs for the same vertex.
- Fix sampling parameter set for GLSL textures (was not setting at all before)
- Fix viewer and unit tests to properly handle either prefiltered or impostance sampling options properly during binding.

### CODE REVIEW

File: source/MaterialXRenderGlsl/GLTextureHandler.cpp [hunk 1]
```diff
@@ -62,6 +62,11 @@ bool GLTextureHandler::acquireImage(const FilePath& filePath,
         _boundTextureLocations.resize(maxTextureUnits, MaterialX::GlslProgram::UNDEFINED_OPENGL_RESOURCE_ID);
     }
 
+    if (filePath.isEmpty())
```
Comment (jstone-lucasfilm) [inline source/MaterialXRenderGlsl/GLTextureHandler.cpp]: Changes to this file look like an unintended reversion of an earlier fix in master.

File: source/MaterialXRenderGlsl/GLTextureHandler.h [hunk 2]
```diff
@@ -43,22 +45,17 @@ class GLTextureHandler : public ImageHandler
     bool createColorImage(const Color4& color,
                           ImageDesc& imageDesc) override;
 
-    /// Acquire an image from the cache or file system.  If the image is not
-    /// found in the cache, then each image loader will be applied in turn.
-    /// @param filePath File path of the image.
-    /// @param imageDesc On success, this image descriptor will be filled out
-    ///    and assigned ownership of a resource buffer.
+    /// Acquire an image from disk.
```
Comment (jstone-lucasfilm) [inline source/MaterialXRenderGlsl/GLTextureHandler.h]: Same as above.

File: source/MaterialXTest/Render.cpp [hunk 3]
```diff
@@ -20,7 +20,7 @@
 #include <MaterialXTest/RenderUtil.h>
 
 #ifdef MATERIALX_BUILD_CONTRIB
-#include <MaterialXContrib/TinyEXRImageLoader.h>
+#include <MaterialXContrib/Handlers/TinyEXRImageLoader.h>
```
Comment (jstone-lucasfilm) [inline source/MaterialXTest/Render.cpp]: Same as above.

File: source/MaterialXView/Material.cpp [hunk 4]
```diff
@@ -369,8 +399,14 @@ void Material::bindImages(mx::GLTextureHandlerPtr imageHandler, const mx::FileSe
         return;
     }
 
+    const std::string IMAGE_SEPERATOR("_");
```
Comment (jstone-lucasfilm) [inline source/MaterialXView/Material.cpp]: Spelling of "separator"

File: source/MaterialXView/Material.cpp [hunk 5]
```diff
@@ -369,8 +399,14 @@ void Material::bindImages(mx::GLTextureHandlerPtr imageHandler, const mx::FileSe
...
+    const std::string UADDRESS_MODE_POST_FIX("_uaddressmode");
+    const std::string VADDRESS_MODE_POST_FIX("_vaddressmode");
+    const std::string FILTER_TYPE_POST_FIX("_filtertype");
+    const std::string DEFAULT_COLOR_POST_FIX("_default");
+    const int INVALID_MAPPED_INT_VALUE = -1; // Any value < 0 is not considered to be invalid
+
     const mx::VariableBlock* publicUniforms = getPublicUniforms();
-    mx::Color4 fallbackColor(0, 0, 0, 1);
```
Comment (jstone-lucasfilm) [inline source/MaterialXView/Material.cpp]: As above, it looks like we've accidentally lost an earlier fix to master in this merge.

File: libraries/pbrlib/genosl/mx_burley_diffuse_brdf.osl [hunk 6]
```diff
@@ -1,7 +1,7 @@
 #include "pbrlib/genosl/lib/mx_fresnel.osl"
 
-void mx_burley_diffuse_brdf(float weight, color reflectance, roughnessinfo roughness, normal N, vector U, output BSDF result)
+void mx_burley_diffuse_brdf(float weight, color reflectance, float roughness, normal N, output BSDF result)
 {
-    // TODO: Implement properly
-    return color * weight * oren_nayar(normal, roughness)
+    // TODO: Implement properly.
+    //result = reflectance * weight * oren_nayar(normal, roughness);
```
[the remark this answers is not in the document]
Comment (bernardkwok) [reply libraries/pbrlib/genosl/mx_burley_diffuse_brdf.osl]: For some reason the built-in oren_nayer cannot be found when compiling for unit tests. We're looking into the reason for this, but for now it's commented out. If we can't figure out why we can uncomment for the PR.

File: source/MaterialXRenderGlsl/GLTextureHandler.cpp (hunk 1 above)
Comment (bernardkwok) [reply source/MaterialXRenderGlsl/GLTextureHandler.cpp]: Looks like we didn't merge this back properly to our repo. Will revert.

File: libraries/pbrlib/genosl/mx_burley_diffuse_brdf.osl (hunk 6 above)
[the remark this answers is not in the document]
Comment (jstone-lucasfilm) [reply libraries/pbrlib/genosl/mx_burley_diffuse_brdf.osl]: That makes sense, and thanks for the clarification.

**Pull request merged**